data/transactions/logic/eval.go

```diff
@@ -272,22 +278,31 @@ type EvalParams struct {
 	// MinTealVersion is nil, we will compute it ourselves
 	MinTealVersion *uint64
 
-	// Amount "overpaid" by the top-level transactions of the group.  Often 0.
-	// When positive, it can be spent by inner transactions.  Shared across a
-	// group's txns, so that it can be updated. nil is interpreted as 0.
+	// Amount "overpaid" by the transactions of the group.  Often 0.  When
+	// positive, it can be spent by inner transactions.  Shared across a group's
+	// txns, so that it can be updated (including upward, by overpaying inner
+	// transactions). nil is treated as 0 (used before fee pooling is enabled).
 	FeeCredit *uint64
 
 	Specials *transactions.SpecialAddresses
 
-	// Total pool of app call budget in a group transaction (nil before pooling enabled)
+	// Total pool of app call budget in a group transaction (nil before budget pooling enabled)
 	PooledApplicationBudget *uint64
 
-	// Total allowable inner txns in a group transaction (nil before pooling enabled)
+	// Total allowable inner txns in a group transaction (nil before inner pooling enabled)
 	pooledAllowedInners *int
 
-	// If non-zero, a txn counter value after which assets and apps should be
-	// allowed w/o foreign array. Left zero until createdResourcesVersion.
-	initialCounter uint64
+	// created contains resources that may be used for "created" - they need not be in
+	// a foreign array. They remain empty until createdResourcesVersion.
+	created *resources
+
+	// Caching these here means the hashes can be shared across the TxnGroup
+	// (and inners, because the cache is shared with the inner EvalParams)
+	appAddrCache map[basics.AppIndex]basics.Address
+
+	// Cache the txid hashing, but do *not* share this into inner EvalParams, as
+	// the key is just the index in the txgroup.
+	txidCache map[int]transactions.Txid
```

Comment:
Bad dog! Review dog is wrong. This field is used, but it is only used through embedding.  I'll change that code to use `cx.EvalParams.txidCache` which should convince it.